### DISCUSSION TITLE
how to edit vim keybindings?

### DISCUSSION THREAD

**Discussion by hopezh** [marimo-team/marimo#3304] [Q&A] [answered]
I use `jk` as `esc` in `neovim`. 

May I ask of there is a way to customize `vim` keybindings in `marimo` to do this?

**Comment by mscolnick** [comment]
Not currently, just the ones you see in the keyboard shortcuts modal
![image](https://github.com/user-attachments/assets/5a737bbf-10ec-48af-aeb0-31b41b836fdc)

**Reply by dmitrydoni** [reply]
Any plans to add this? It's a very common key binding among vim users.

**Reply by mscolnick** [reply]
@dmitrydoni, can you share how you'd configure this? i assume it would be scoped to the vim mode?

**Reply by jabellcu** [reply]
Perhaps it could be configured just like standard VIm, with a .vimrc file? There could be a parameter to point at .vimrc location.

**Reply by dmitrydoni** [reply]
@mscolnick Yes, it is a key to exit the Insert mode and enter Normal mode. 

When Vim is enabled as a top-level setting in Marimo, it could follow the same user-defined patterns for each mode. Key mappings is one of those. Ideally, it would support Lua like Nvim, but having at least the old way with `.vimrc` would be great as suggested above (thanks @jabellcu !)

For example, in Obsidian, they do it this way, so for key binding there is a file `.obsidian.vimrc` with  content like this:
```
" Exit Insert mode
imap jk <Esc> 
```

**Reply by mscolnick** [reply]
a `vimrc` (either path to one or the contents) as marimo config makes sense to me.

there is some prior art for parsing and adding extensions in: https://github.com/esm7/obsidian-vimrc-support/blob/47de57947131d56d06cf433cf49390e493290439/main.ts

and docs for adding custom map keys: https://github.com/replit/codemirror-vim#map-keys

would anyone be open to making the change / contribution? I am personally not a vim user (heavy multi-cursor!) so would not be as effective

**Reply by andyrift** [reply]
I am a neovim user an I just started using marimo.

As far as I understand it would be a whole lot simpler at the moment to just add support for some vimrc commands than to develop a programmatic approach with a normal language like in neovim? Like with python or with lua?
And that is because codemirror-vim already supports adding and removing any keymaps, and vimrc in its simple form is easy to parse?

Would it be helpful if I tried to figure out how all this works? I have some prior experience with prosemirror (not codemirror but also from marijnh).

I personally am having issues with :w seemingly not doing anything (I want it to trigger saving the notebook) and I can't find any documentation on commands in vim mode, but maybe all that is beside the point.

**Reply by mscolnick** [reply]
@andyrift, we would love some help on this. i don't use `vim` personally, but if you can help write the typescript in the format that `codemirror-vim` expects, I can try to 1) wire the vimrc from the file to the frontend and 2) verify it works

additionally `:w` does nothing as of now, we but can map this to save

**Reply by andyrift** [reply]
@mscolnick, I played a little with the code and I managed to hardcode some mappings with `Vim.map`, `Vim.defineEx` and other commands from there.
Even this works:
``` typescript
Vim.defineEx("write", "w", () => { console.log("Save") })
Vim.noremap("S", ":w<CR>", "normal");
```
(pressing `Shift+s` in normal mode executes the `:w` command which itself logs the `"Save"` message)

I think I see how to allow this with vimrc, but for now only the `map`  and `noremap` commands with their `i`, `n` and `v` variants (or only these variants).
Almost no parsing is required, as I see it.

And the rest of the vimrc commands might be ignored or yelled at.

**Reply by mscolnick** [reply]
thats great. let me start the two PRs. one for write and the other for `vimrc` wiring

**Reply by andyrift** [reply]
Do we need as true to the original vimrc parsing and command behavior, or "barely works, able to use basic mappings" would be enough for now?

I have a working prototype.
Hardcoded string resembling vimrc is used to set the keymaps that then work in the editor. Like this:
``` Typescript
const vimrc = `
nmap W :w<CR>
inoremap jk <Esc>
`;
```

**Reply by mscolnick** [reply]
@andyrift - i may have a solution here. i tests with some basic commands but mind checking my parsing logic? https://github.com/marimo-team/marimo/pull/4473

**Comment by andyrift** [comment, marked as answer]
It is now possible to supply your own vimrc with mappings, got added in marimo 0.12.9 (#4473, #4487)
